README
Test::Deep::Filter

This module exists so that one can apply some kind of transform of the
content of some node, before applying a deep testing structure against
it.

This is especially the kind of domain Test::Deep::YAML and
Test::Deep::JSON are targeted at, but implemented in a more generic way.

  cmp_deeply( $got, { payload => filter(\&decode_json, { x => 1 }) } )

This would perform matching against a "Perl" data structure called $got,
and find its key called "payload", and, if existing, filter the key via
&decode_json, and then compare the resulting structure with "{ x => 1
}".

This would in theory be equivalent to:

  cmp_deeply( $got, { payload => json({ x => 1 }) } )

Except may be tailored for what ever local decoding scheme you may
require, e.g.:

  cmp_deeply( $got, {
      payload => filter(sub {
        my ( $got ) = @_;
        return parse_ini( $got );
      }, { x => 1 })
  });

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Test-Deep-Filter

0.001002 2017-03-07T10:58:37Z a829f0c

0.001001 2017-03-07T10:57:32Z b387f9f
 - Bugfix: Avoid test failures under -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: +3 ↑2 (suggests: ↑2)
 - test: ↓1

0.001000 2015-08-14T10:04:05Z a46add0
 - First version.


lib/Test/Deep/Filter.pm
use 5.006;    # our
use strict;
use warnings;

package Test::Deep::Filter;

our $VERSION = '0.001002';

# ABSTRACT: Perform a filter on a matched element before doing sub-matching

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

require Exporter;
*import = \&Exporter::import;    ## no critic (ProhibitCallsToUnexportedSubs)

our @EXPORT_OK = ('filter');

sub filter {
  my ( $code, $expected ) = @_;
  require Test::Deep::Filter::Object;
  return Test::Deep::Filter::Object->new( $code, $expected );
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Deep::Filter - Perform a filter on a matched element before doing sub-matching

=head1 VERSION

version 0.001002

=head1 DESCRIPTION

This module exists so that one can apply some kind of transform of the content
of some node, before applying a deep testing structure against it.

This is especially the kind of domain Test::Deep::YAML and Test::Deep::JSON are
targeted at, but implemented in a more generic way.

  cmp_deeply( $got, { payload => filter(\&decode_json, { x => 1 }) } )

This would perform matching against a C<Perl> data structure called C<$got>,
and find its key called C<payload>, and, if existing, filter the key via
C<&decode_json>, and then compare the resulting structure with
C<< { x => 1 } >>.

This would in theory be equivalent to:

  cmp_deeply( $got, { payload => json({ x => 1 }) } )

Except may be tailored for what ever local decoding scheme you may require,
e.g.:

  cmp_deeply( $got, {
      payload => filter(sub {
        my ( $got ) = @_;
        return parse_ini( $got );
      }, { x => 1 })
  });

=head1 FUNCTIONS

=head2 C<filter>

  my $object = filter( \&filter_function, $expected );

  cmp_deeply( $got, filter( \&filter_function, $expected ));

A C<filter> can take any value into C<&filter_function>, and return any value
and the C<Test::Deep> infrastructure will handle the rest.

If the value cannot be parsed, or is of a type you cannot handle, or the
internal logic returns an invalid value ( including any failures in internal
logic that fail to parse, etc) then you must cause the function to throw an
exception with "die", and the nested C<$expected> will be assumed to be false.

The "die" message will be used as a diagnostic for why the match failed.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Test/Deep/Filter/Object.pm
use 5.006;    # our
use strict;
use warnings;

package Test::Deep::Filter::Object;

# ABSTRACT: Internal plumbing for Test::Deep::Filter

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

our $VERSION = '0.001002';

use parent 'Test::Deep::Cmp';







sub init {
  my ( $self, $filter, $expected ) = @_;
  $self->{ __PACKAGE__ . '_filter' }   = $filter;
  $self->{ __PACKAGE__ . '_expected' } = $expected;
  return;
}











## no critic ( RequireArgUnpacking, RequireFinalReturn )
sub filter   { $_[0]->{ __PACKAGE__ . '_filter' } }
sub expected { $_[0]->{ __PACKAGE__ . '_expected' } }
## use critic







sub descend {
  my ( $self, $got ) = @_;
  delete $self->{ __PACKAGE__ . '_error' };
  my $return;
  {
    local $@ = undef;
    $return = eval { local $_ = $got; $self->filter->($got) };
    if ( defined $@ and length $@ ) {
      $self->{ __PACKAGE__ . '_error' } = $@;
      return 0;
    }
  }
  require Test::Deep;
  return Test::Deep::wrap( $self->expected )->descend($return);    ## no critic (ProhibitCallsToUnexportedSubs)
}







sub diagnostics {
  my ( $self, $where, $last_exp ) = @_;
  return $self->{ __PACKAGE__ . '_error' } if exists $self->{ __PACKAGE__ . '_error' };
  return $self->expected->diagnostics( $where, $last_exp );
}
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Deep::Filter::Object - Internal plumbing for Test::Deep::Filter

=head1 VERSION

version 0.001002

=head1 METHODS

=head2 C<init>

  my $object = Test::Deep::Filter::Object->new( $filter, $expected_structure );

=head2 C<filter>

  my $filter_sub = $object->filter;

=head2 C<expected>

  my $expected = $object->expected;

=head2 C<descend>

  my $result = $object->descend( $got );

=head2 C<diagnostics>

  my $diagnostics = $object->diagnostics($where, $last_exp);

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perform a filter on a matched element before doing sub-matching",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-Deep-Filter",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::Deep::Filter",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Test::Deep" => 0,
    "Test::Deep::Cmp" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "Test::Tester" => "0.08"
  },
  "VERSION" => "0.001002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::Deep" => 0,
  "Test::Deep::Cmp" => 0,
  "Test::More" => 0,
  "Test::Tester" => "0.08",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
